src/ApiCore/GapicClientTrait.php

```diff
@@ -0,0 +1,316 @@
+ * met:
+ *
+ *     * Redistributions of source code must retain the above copyright
+ * notice, this list of conditions and the following disclaimer.
+ *     * Redistributions in binary form must reproduce the above
+ * copyright notice, this list of conditions and the following disclaimer
+ * in the documentation and/or other materials provided with the
+ * distribution.
+ *     * Neither the name of Google Inc. nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+namespace Google\ApiCore;
+
+use Google\ApiCore\LongRunning\OperationsClient;
+use Google\ApiCore\Middleware\AgentHeaderMiddleware;
+use Google\ApiCore\Middleware\RetryMiddleware;
+use Google\ApiCore\Transport\TransportInterface;
+use Google\Protobuf\Internal\Message;
+
+/**
+ * Common functions used to work with various clients.
+ */
+trait GapicClientTrait
+{
+    use ArrayTrait;
+    use ValidationTrait;
+
+    private static $gapicVersion;
+    private $retrySettings;
+    private $serviceName;
+    private $agentHeaderDescriptor;
+    private $descriptors;
+    private $transport;
+    private $transportCallMethods = [
+        Call::UNARY_CALL => 'startUnaryCall',
+        Call::BIDI_STREAMING_CALL => 'startBidiStreamingCall',
+        Call::CLIENT_STREAMING_CALL => 'startClientStreamingCall',
+        Call::SERVER_STREAMING_CALL => 'startServerStreamingCall',
+    ];
+
+    private static function getGapicVersion(array $options)
+    {
+        if (!self::$gapicVersion) {
+            if (isset($options['versionFile'])) {
+                if (file_exists($versionFile)) {
+                    self::$gapicVersion = trim(file_get_contents(
+                        $options['versionFile']
+                    ));
+                }
+            } elseif (isset($options['libVersion'])) {
+                $options['gapicVersion'] = $options['libVersion'];
+            }
+        }
+
+        return self::$gapicVersion;
+    }
+
+    /**
+     * Configures the GAPIC client based on an array of options.
+     *
+     * @param array $options {
```

Comment:
Ae these docs out of date? Looks like we check for a lot of options not listed

Reply:
Fixed by David.